sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala

```diff
@@ -0,0 +1,384 @@
+    // cannot reuse the same code for the other types.
+    // Overflow happens if the arguments have different signs and the sign of the result is
+    // different from the sign of subtractend (RHS).
+    val numRows = ret.getRowCount.toInt
+    val zero = BigDecimal(0).bigDecimal
+    val overflow = withResource(DecimalUtils.lessThan(rhs, zero, numRows)) { rhsLz =>
+      val argsSignDifferent = withResource(DecimalUtils.lessThan(lhs, zero, numRows)) { lhsLz =>
+        lhsLz.notEqualTo(rhsLz)
+      }
+      withResource(argsSignDifferent) { argsSignDifferent =>
+        val resultAndSubtrahendSameSign =
+          withResource(DecimalUtils.lessThan(ret, zero)) { resultLz =>
+            rhsLz.equalTo(resultLz)
+          }
+        withResource(resultAndSubtrahendSameSign) { resultAndSubtrahendSameSign =>
+          resultAndSubtrahendSameSign.and(argsSignDifferent)
+        }
+      }
+    }
+    withResource(overflow) { overflow =>
+      if (failOnError) {
+        withResource(overflow.any()) { any =>
+          if (any.isValid && any.getBoolean) {
+            throw new ArithmeticException("One or more rows overflow for Subtract operation.")
+          }
+        }
+        ret.incRefCount()
+      } else {
+        withResource(GpuScalar.from(null, dataType)) { nullVal =>
+          overflow.ifElse(nullVal, ret)
+        }
+      }
+    }
+  }
+
+  override def doColumnar(lhs: BinaryOperable, rhs: BinaryOperable): ColumnVector = {
+    val ret = super.doColumnar(lhs, rhs)
+    withResource(ret) { ret =>
+      // No shims are needed, because it actually supports ANSI mode from Spark v3.0.1.
+      if (failOnError && GpuAnsi.needBasicOpOverflowCheck(dataType) ||
+          GpuTypeShims.isSupportedDayTimeType(dataType) ||
+          GpuTypeShims.isSupportedYearMonthType(dataType)) {
+        // For day time interval, Spark throws an exception when overflow,
+        // regardless of whether `SQLConf.get.ansiEnabled` is true or false
+        basicOpOverflowCheck(lhs, rhs, ret)
+      }
+
+      if (dataType.isInstanceOf[DecimalType]) {
+        decimalOpOverflowCheck(lhs, rhs, ret)
+      } else {
+        ret.incRefCount()
+      }
+    }
+  }
+
+  override def resultDecimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType = {
+    val resultScale = max(s1, s2)
+    val resultPrecision = max(p1 - s1, p2 - s2) + resultScale + 1
+    if (allowPrecisionLoss) {
+      DecimalType.adjustPrecisionScale(resultPrecision, resultScale)
+    } else {
+      DecimalType.bounded(resultPrecision, resultScale)
+    }
+  }
+}
+
+trait GpuDivModLikeWithPromote extends GpuDivModLike with DecimalWithPromote
+
+trait DecimalWithPromote extends CudfBinaryOperator {
+  private def getCastedVectorIfNeeded(
+      vector: GpuColumnVector,
+      d: DecimalType, c: DType): GpuColumnVector = {
+    if (!vector.dataType.sameType(d)) {
+      GpuColumnVector.from(vector.getBase.castTo(c), d)
```

Comment:
I still don't understand why we are doing this for integral divide?

Reply:
As I understand the code in IntegralDivide didn't have to promote precision as it was being done in `DecimalPrecision.scala` [here](https://github.com/apache/spark/commit/301a13963808d1ad44be5cacf0a20f65b853d5a2?diff=unified#diff-a3c51d44e59e904c5ac9f75b2abc49959181145649dcc51c2437f7fbc7bb5be7L186). Now if we don't do this, the resultDecimalType will return a DecimalType that may not match the type of the inputs. Hence we have to cast them to the expected type. Otherwise, the implicit conversion fails, now that I think about this again, we could try to override the outputType. Let me try that

Reply:
Yeah, that doesn't work either as that is just deciding what the outputType should be, the inputs still have to be casted.

Reply:
My main concern is that we get the correct answer.  From what I have seen the new Integral divide code for decimals does the divide using the original values, but the output scale is always set to 0, no matter what the precision is. Then the result is cast to a long. I don't think that corresponds to what this patch is doing. 

For example what happens if we do `1 DIV 0.5` The answer should be `2`. But what happens if the decimal types make it so we lose information if we do the cast ahead of time?  `CAST(1 AS DECIMAL(38, 0)) DIV CAST(0.5 AS DECIMAL(38, 38))`. The answer on older versions of Spark is null because the cast of 1 to decimal(38,38) overflowed.  Your new code does not have the overflow checking. I am not sure what the cast is going to do here, and I don't know what the answer is going to be. I think in 3.4 the answer will still be 2 because it does the divide prior to any other operations, and then figures out how to make it a long.

Reply:
I have repurposed the `DecimalDivide` and used it for `IntegralDivide` I've had to modify the divide128 kernel for it as the regular divide rounds up the quotient.You can look at the cudf PR [here](https://github.com/NVIDIA/spark-rapids-jni/pull/803)